File: backends/fastapi/app/services/capsule_recommendation_service.py
"""为创建页提供 AI 推荐的胶囊主题列表（时间跨度兼顾近远）。"""

from __future__ import annotations

import logging

from app.core.config import REPO_ROOT, settings
from app.schemas.capsule_recommendation import (
    CapsuleRecommendationItem,
    CapsuleRecommendationList,
)
from app.services.llm_client import LlmClientError, generate_capsule_recommendations

logger = logging.getLogger(__name__)

_MIN_ITEMS = 3
_MAX_ITEMS = 8


def _prompt_template() -> str:
    path = REPO_ROOT / "spec" / "llm" / "capsule-recommendation.prompt.md"
    if path.exists():
        return path.read_text(encoding="utf-8")
    return (
        "你是中文写作助手。请生成 {COUNT} 条互不重复的时光胶囊主题推荐，时间跨度兼顾近远。"
        "每条含 title（1~24 字中文标题）、hint（一句话灵感）、openInDays（1~3650 整数）。"
        '只返回严格 JSON：{"items":[{"title":"...","hint":"...","openInDays":30}]}。'
    )


def _build_prompt(count: int) -> str:
    return _prompt_template().replace("{COUNT}", str(count))


def _clamp_days(raw: object) -> int | None:
    try:
        if isinstance(raw, bool):
            return None
        if isinstance(raw, int):
            days = raw
        elif isinstance(raw, float):
            days = int(raw)
        elif isinstance(raw, str):
            days = int(raw.strip())
        else:
            return None
    except (ValueError, TypeError):
        return None
    return max(1, min(3650, days))


def _clean(text: object, limit: int) -> str:
    cleaned = str(text).strip().replace("\n", " ").replace("\r", " ")
    cleaned = cleaned.strip("#*` 　\"'《》【】")
    return cleaned[:limit].strip()


def _parse_items(raw_items: object) -> list[CapsuleRecommendationItem]:
    items: list[CapsuleRecommendationItem] = []
    seen_titles: set[str] = set()
    if not isinstance(raw_items, list):
        return items
    for raw in raw_items:
        if not isinstance(raw, dict):
            continue
        title = _clean(raw.get("title", ""), 60)
        hint = _clean(raw.get("hint", ""), 80)
        days = _clamp_days(raw.get("openInDays"))
        if not title or not hint or days is None:
            continue
        if title in seen_titles:
            continue
        seen_titles.add(title)
        items.append(CapsuleRecommendationItem(title=title, hint=hint, openInDays=days))
    return items


def get_capsule_recommendations(count: int, locale: str) -> CapsuleRecommendationList:
    """推荐为锦上添花：LLM 不可用时返回空列表，不做本地兜底、不报错。"""
    count = max(_MIN_ITEMS, min(_MAX_ITEMS, count))
    items: list[CapsuleRecommendationItem] = []

    try:
        result = generate_capsule_recommendations(_build_prompt(count))
        items = _parse_items(result.get("items"))[:count]
    except (KeyError, TypeError, ValueError, LlmClientError) as e:
        logger.info("Capsule recommendations unavailable; returning empty list: %s", e)

    generated_by = f"{settings.llm_provider}:{settings.llm_model}" if items else "none"
    return CapsuleRecommendationList(items=items, generatedBy=generated_by, cached=False)

File: backends/nest/src/capsule-recommendation/capsule-recommendation.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import * as fs from 'fs';
import * as path from 'path';
import { AppConfig } from '../config/configuration';
import { LlmClientService } from '../llm/llm-client.service';

const MIN_ITEMS = 3;
const MAX_ITEMS = 8;

export interface RecommendationItem {
  title: string;
  hint: string;
  openInDays: number;
}

export interface RecommendationList {
  items: RecommendationItem[];
  generatedBy: string;
  cached: boolean;
}

@Injectable()
export class CapsuleRecommendationService {
  private readonly logger = new Logger(CapsuleRecommendationService.name);

  constructor(
    private readonly config: ConfigService<AppConfig>,
    private readonly llm: LlmClientService,
  ) {}

  private promptTemplate(): string {
    const repoRoot = this.config.get('repoRoot', { infer: true }) ?? process.cwd();
    const promptPath = path.join(repoRoot, 'spec', 'llm', 'capsule-recommendation.prompt.md');
    try {
      return fs.readFileSync(promptPath, 'utf-8');
    } catch {
      return (
        '你是中文写作助手。请生成 {COUNT} 条互不重复的时光胶囊主题推荐，时间跨度兼顾近远。' +
        '每条含 title（1~24 字中文标题）、hint（一句话灵感）、openInDays（1~3650 整数）。' +
        '只返回严格 JSON：{"items":[{"title":"...","hint":"...","openInDays":30}]}。'
      );
    }
  }

  private buildPrompt(count: number): string {
    return this.promptTemplate().replace('{COUNT}', String(count));
  }

  private clampDays(raw: unknown): number | null {
    let days: number;
    if (typeof raw === 'boolean') return null;
    if (typeof raw === 'number') days = Math.round(raw);
    else if (typeof raw === 'string') days = parseInt(raw, 10);
    else return null;
    if (isNaN(days)) return null;
    return Math.min(3650, Math.max(1, days));
  }

  private clean(text: unknown, limit: number): string {
    let cleaned = String(text ?? '').trim().replace(/[\r\n]+/g, ' ');
    cleaned = cleaned.replace(/^[#*`　 "'《》【】]+|[#*`　 "'《》【】]+$/g, '').trim();
    return [...cleaned].slice(0, limit).join('');
  }

  private parseItems(raw: unknown): RecommendationItem[] {
    const items: RecommendationItem[] = [];
    if (!Array.isArray(raw)) return items;
    const seen = new Set<string>();
    for (const entry of raw) {
      if (typeof entry !== 'object' || entry === null) continue;
      const e = entry as Record<string, unknown>;
      const title = this.clean(e.title, 60);
      const hint = this.clean(e.hint, 80);
      const days = this.clampDays(e.openInDays);
      if (!title || !hint || days === null) continue;
      if (seen.has(title)) continue;
      seen.add(title);
      items.push({ title, hint, openInDays: days });
    }
    return items;
  }

  /** 推荐为锦上添花：LLM 不可用时返回空列表，不做本地兜底、不报错。 */
  async getRecommendations(count: number, _locale: string): Promise<RecommendationList> {
    const n = Math.min(MAX_ITEMS, Math.max(MIN_ITEMS, count));
    let items: RecommendationItem[] = [];
    try {
      const result = await this.llm.generateRecommendations(this.buildPrompt(n));
      items = this.parseItems(result.items).slice(0, n);
    } catch (e) {
      this.logger.log(`Capsule recommendations unavailable; returning empty list: ${e}`);
    }

    const provider = this.config.get('llmProvider', { infer: true }) ?? 'openai';
    const model = this.config.get('llmModel', { infer: true }) ?? 'gpt-4.1-mini';
    return {
      items,
      generatedBy: items.length > 0 ? `${provider}:${model}` : 'none',
      cached: false,
    };
  }
}
